jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -715,26 +715,19 @@ public void run()
         @Override
         public void succeeded()
         {
-            HttpStream stream;
-            boolean completeStream;
-            try (AutoLock ignored = _lock.lock())
-            {
-                assert _callbackCompleted;
-                assert _callbackFailure == null;
-                _streamSendState = StreamSendState.LAST_COMPLETE;
-                completeStream = _handling == null;
-                stream = _stream;
-            }
-
-            if (completeStream)
-                completeStream(stream, null);
+            complete(null);
         }
 
         /**
          * Called only as {@link Callback} by last send from {@link ChannelCallback#succeeded}
          */
         @Override
         public void failed(Throwable failure)
+        {
+            complete(failure);
+        }
+
+        private void complete(Throwable failure)
```

Comment:
Any chance you can put a comment on the HandlerInvoker class to say how/when/why it is used as a Callback... as I'm getting confused with the _callbackCompleted below, which is a different callback I think (I know it is kind of orthogonal to this)

Other than that, this looks good.